### PR TITLE
[C API] Add static linking for WIN32

### PR DESCRIPTION
**Pull request by taniabogatsch**
Follow-up of https://github.com/duckdb/duckdb/pull/12709 integrating the changes of @JAicewizard into the new C API generation scripts.

### CODE REVIEW

File: src/include/duckdb.h [hunk 1]
```diff
@@ -29,11 +33,15 @@
 //! duplicate of duckdb/main/winapi.hpp
 #ifndef DUCKDB_EXTENSION_API
 #ifdef _WIN32
+#ifdef DUCKDB_STATIC
+#define DUCKDB_API
```
Comment (samansmink) [inline src/include/duckdb.h:37]: shouldn't this be `DUCKDB_EXTENSION_API`?

**Comment by taniabogatsch** [comment]
Initial discussion here: https://github.com/duckdb/duckdb/discussions/2774

File: src/include/duckdb.h [hunk 2]
```diff
@@ -16,11 +16,15 @@
 //! duplicate of duckdb/main/winapi.hpp
 #ifndef DUCKDB_API
 #ifdef _WIN32
+#ifdef DUCKDB_STATIC
```
Comment (samansmink) [inline src/include/duckdb.h:19]: Maybe this define should be a bit more descriptive?

File: src/include/duckdb.h (hunk 1 above)
Comment (taniabogatsch) [reply src/include/duckdb.h:37]: Oh, yes, thanks for catching this!

File: src/include/duckdb.h (hunk 2 above)
Comment (samansmink) [reply src/include/duckdb.h:19]: not sure what else though, so maybe its fine

**Comment by JAicewizard** [comment]
I already made a PR with these changes (#12709), should I close that one? this one seems better. But this also solves #2774

File: src/include/duckdb.h (hunk 2 above)
Comment (taniabogatsch) [reply src/include/duckdb.h:19]: I'll push `DUCKDB_STATIC_BUILD`?

File: src/include/duckdb/common/winapi.hpp [hunk 3]
```diff
@@ -9,24 +9,32 @@
 #pragma once
 
 #ifndef DUCKDB_API
-#if defined(_WIN32) && !defined(__MINGW32__)
```
[the remark this answers is not in the document]
Comment (JAicewizard) [reply src/include/duckdb/common/winapi.hpp:12]: I think it should go, this inconsistency between duckdb.h and winapi.hpp seems like a bug to me

[the remark this answers is not in the document]
Comment (samansmink) [reply src/include/duckdb/common/winapi.hpp:12]: I'm not 100% sure but i would agree with @JAicewizard that that sounds likely

[the remark this answers is not in the document]
Comment (taniabogatsch) [reply src/include/duckdb/common/winapi.hpp:12]: Alright, thanks!

**Comment by taniabogatsch** [comment]
The MIngW build is failing; I'll have a look tomorrow.
https://github.com/duckdb/duckdb/actions/runs/10703802188/job/29678308327?pr=13744

**Comment by taniabogatsch** [comment]
Okay, somehow we need `#if defined(_WIN32) && !defined(__MINGW32__)` for the `Windows / MingW (64 Bit)` build to succeed. Unsure why, but this PR does not seem like the place to find out.

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**